### DISCUSSION TITLE
Setting a “standard” message transmission rate.

### DISCUSSION THREAD

**Discussion by joehodgy** [collin80/SavvyCAN#303] [General]
More of a coding question than a SavvyCan question and for that I apologise!

This relates to replacing an existing OEM control module (or module) on the can with a standalone tool such as the Macchina M2.

The modules, at least the one I’m messing with right now, on this car send their data at relatively fixed intervals - around 60ms. I want to replicate this as far as possible to keep bus availability high and to avoid any issues with time-out or missing messages as other modules wait for that data.

Can anyone provide example code that would allow this? The sendFrame seems to send as soon as there is availability, whenever there is availability and floods the bus.

Thanks all
Joe

(I asked on the macchina forum but no response.)

**Comment by mitchdetailed** [comment]
it'll vary depending on the hardware, but i tend to use socketcan and
python-can as a primary source of receiving and transmitting

**Comment by collin80** [comment]
Yes, sendFrame is immediate (more or less). What you need to do is detect when enough time has passed and send the frame when you're ready. There are two basic options:

1. You can read the milliseconds since boot and trigger every 60ms. Something like

`uint32_t nextTick = 60`

`void loop()`
`{`
`   if (millis() + 60 > nextTick) `
`   {`
`      nextTick = millis() + 60;`
`      Can0.sendFrame(yourFrame);`
`   }`
`}`

2. You can set up a hardware timer to interrupt every 60ms then either send directly from the interrupt handler (which you probably should not do) or set a boolean flag and detect it in your loop function. This has many more options and I'm not going to include a full example here but you can find info on hardware timers online. Arguably, this might be the more "correct"  option but quite often I use option 1 above because it's simple and easy and you can very quickly change the timing or even set the next tick to different intervals really easily.

**Reply by joehodgy** [reply]
Thanks Collin :) really appreciate this isn't really what the discussion board is for!

I've coded it using a millis timer but think I'll go back and re-do it your way for 'correctness'/simplicity and easiness XD

My next step is coding RPM and road speed from exterior sources to can, thanks again

Joe
